usbd: device_next: cdc: add option to enable short packet terminate

zeonchew opened the pull request:
For windows default USB CDC host driver for windows, `usbser.sys`, it is expected that USB device always indicate completion of transmission by short packet. In the case where the last packet length is multiple of max packet size of the BULK IN endpoint, the USB device shall indicate completion of transmission by sending a zero-length-packet. In case where transmission completion indication is missing, the driver might withhold the data from application that is listening to the COM port until more data is transmit to the host. 

This commit adds an option to `usbd_cdc_acm` driver to enable the sending of zero length packet mentioned above, hence ensuring the condition of short packet terminate is fulfilled. This option is currently set as disabled by default in this PR.

zeonchew committed:
usbd: device_next: cdc: add option to enable short packet terminate

For usbser.sys driver, which is the default USB host driver for CDC
devices in windows, it is expected that USB device always indicate
completion of transmission by short packet. In case where the last
packet length is multiple of max packet size of the BULK IN endpoint,
the USB device shall indicate completion of transmission by sending a
zero length packet. This commit adds the sending of the short packet
termination mentioned above, ensuring condition of short packet
terminate is fulfilled.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

tmon-nordic reviewed (changes requested):
Don't use #ifdefs. Replace them with `if (IS_ENABLED(CONFIG_...))` that always make the compiler check the code regardless of the KConfig option value and simply eliminate never taken branches (if KConfig is disabled).

zeonchew commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -111,6 +111,12 @@ struct cdc_acm_uart_data {
 	atomic_t state;
 	struct k_sem notif_sem;
+#ifdef CONFIG_USBD_CDC_ACM_SHORT_PACKET_TERMINATE_ENABLED
+	/* size of packet last tx */
+	uint32_t last_tx_size;
+	/* udc enumerated speed */
+	uint32_t ep_mps_mask;
```
Hi @tmon-nordic, ep_mps_mask was wrongly described here, I apologize for the mistake. The ep_mps_mask is saved so that we do not need to query for device speed and calculate the mask every time we want to check whether a transaction requested is actually multiple of EP's Max-Packet-Size. 

I totally agree that we only need a boolean to keep "needs ZLP" information, in case that there isn't more data pushed down to tx_fifo by application layer. will modify that soon and push it up for your review. 

Thanks @tmon-nordic

zeonchew replied:
Hi @tmon-nordic, after recent changes, I think this is no longer relevant. Shall we resolve this comment? Thanks 😉

zeonchew commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -203,6 +209,16 @@ static int usbd_cdc_acm_request(struct usbd_class_data *const c_data,
+
+	if ((data->ep_mps_mask == 0) && (udc_api->device_speed)) {
+		data->ep_mps_mask = (udc_api->device_speed(uds_ctx->dev) == UDC_BUS_SPEED_FS)
+					    ? data->desc->if1_in_ep.wMaxPacketSize
+					    : data->desc->if1_hs_in_ep.wMaxPacketSize;
```
Hi @tmon-nordic , Thanks for pointing out. I agree with you. In fact I changed to this "descriptor way" just before I created the PR. I made a mistake here as well that the mask should be maxPacketSize - 1 instead. Will change it back to numbers as suggested soon and push the commit up for your review. 

Thanks again @tmon-nordic

zeonchew replied:
Hi @tmon-nordic, after recent changes, I think this is no longer relevant. Shall we resolve this comment? Thanks 😉

zeonchew commented:
> Don't use #ifdefs. Replace them with `if (IS_ENABLED(CONFIG_...))` that always make the compiler check the code regardless of the KConfig option value and simply eliminate never taken branches (if KConfig is disabled).

Thanks for the guidance. Will change accordingly soon and push the commit up for your review.

zeonchew commented:
Hi @tmon-nordic, I have addressed the review comments above in [Commit 68d18a23742](https://github.com/zephyrproject-rtos/zephyr/pull/79193/commits/68d18a23742f839970b0df2efd07f01026a067c6 ). Please kindly help to review whether the changes looks good to you. Thank you.

zeonchew commented:
> The PR should contain the "final version" of the commits. In this PR you should squash the "review comments" commit together with the original commit.

Hi @tmon-nordic, thanks for the guidance, I have squashed the commits together and force pushed it up.

jfischer-no commented on subsys/usb/device_next/class/Kconfig.cdc_acm:
```diff
@@ -21,6 +21,15 @@ config USBD_CDC_ACM_STACK_SIZE
 	help
 	  USB CDC ACM workqueue stack size.
 
+config USBD_CDC_ACM_SHORT_PACKET_TERMINATE_ENABLED
```
Please come up with a concise option name.

zeonchew replied:
> Please come up with a concise option name.

Hi @jfischer-no, 

The naming "SHORT_PACKET_TERMINATE" was taken from WinUSB policy name [as documented here](https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/winusb-functions-for-pipe-policy-modification). 

How about naming it `USBD_CDC_ACM_ENSURE_SHORT_PKT_FOR_LAST_TX_PKT`, which implies more directly that the option ensures that the last TX packet will be a short packet. 

What do you think?

zeonchew replied:
I have changed it to `USBD_CDC_ACM_ENSURE_SHORT_PKT_FOR_LAST_TX_PKT` in the latest commit push for now as I couldn't think of other better name for now.  Please let me know if you have any concern on the name.

jfischer-no replied:
The new name is the opposite of concise. No?

zeonchew replied:
My bad, understand it wrongly for being not precise enough, but it seems like what you mean is to make it shorter but with the same meaning. What do you think about `USB_CDC_ACM_TX_ZLP_TERMINATE_ENABLED` ?

zeonchew replied:
Hi @jfischer-no, changed the config name to `USBD_CDC_ACM_TX_ZLP_ENABLED` and pushed the change. Please review, thank you.

jfischer-no commented on subsys/usb/device_next/class/Kconfig.cdc_acm:
```diff
@@ -21,6 +21,15 @@ config USBD_CDC_ACM_STACK_SIZE
+	bool "USB CDC ACM transmission always end with short packet"
+	default n
+	help
+	  USB CDC ACM driver sends a zero-length-packet to USB host
+	  following the last packet transmitted, if its length is
```
There is not "USB CDC ACM driver".

zeonchew replied:
Hi @jfischer-no , 
I have modified it according to latest context, and stop calling USB CDC ACM a driver as well: 
https://github.com/zephyrproject-rtos/zephyr/blob/c0d442e9c3ed9f63237d46015179f109993a21b0/subsys/usb/device_next/class/Kconfig.cdc_acm#L24-L33
Please help to review and see whether its all good 😉

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -202,6 +206,15 @@ static int usbd_cdc_acm_request(struct usbd_class_data *const c_data,
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_SHORT_PACKET_TERMINATE_ENABLED)) {
+		if ((data->ep_mps_mask == 0) && (udc_api->device_speed)) {
+			data->ep_mps_mask =
+				(udc_api->device_speed(uds_ctx->dev) == UDC_BUS_SPEED_FS) ? 0x3F
+											  : 0x1FF;
```
Please familiarize yourself with the code base first.

zeonchew replied:
Hi @jfischer-no, 
Thanks for the pointers below, let me study the existing zlp handling and get back with a handling that matches the existing design.  Thanks again 😉

zeonchew replied:
Hi @jfischer-no, 
I have changed it to similar handling as what you have pointed out in CDC ECM. Please help to have a look. 😄

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -549,6 +566,10 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 	len = ring_buf_get(data->tx_fifo.rb, buf->data, buf->size);
 	net_buf_add(buf, len);
 
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_SHORT_PACKET_TERMINATE_ENABLED)) {
+		data->zlp_needed = (len && ((len & data->ep_mps_mask) == 0));
+	}
+
```
https://github.com/zephyrproject-rtos/zephyr/blob/9c0f92db475f0e27673ac366276026eddf1a150c/subsys/usb/device_next/class/usbd_cdc_ecm.c#L522-L524

zeonchew replied:
Hi @jfischer-no  Thanks for the pointer. I have adopted this in this PR. 👍

jfischer-no commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -549,6 +566,10 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 	net_buf_add(buf, len);
 
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_SHORT_PACKET_TERMINATE_ENABLED)) {
+		data->zlp_needed = (len && ((len & data->ep_mps_mask) == 0));
```
All unnecessary parentheses cost one ounce of gold each.

jfischer-no reviewed (changes requested):
>For some USB host driver

What host driver?

zeonchew commented:
> For windows default USB CDC host driver for windows, `usbser.sys`,

Hi @jfischer-no, for now the driver that I know driver to require short packet termination as mentioned is `usbser.sys` driver, which is the default driver for Windows for USB CDC devices. It was written that way to point out possibility of other USB host driver that might exhibit the same requirement.
However, I do agree your view on making the description more concise to what is observed and what we are doing to resolve it. I have modified the PR description as you suggested.

ycsin commented on subsys/usb/device_next/class/Kconfig.cdc_acm:
```diff
@@ -21,6 +21,16 @@ config USBD_CDC_ACM_STACK_SIZE
 
+config USBD_CDC_ACM_ENSURE_SHORT_PKT_FOR_LAST_TX_PKT
+	bool "USB CDC ACM transmission always end with short packet"
+	default n
```
`n` should be implied
```suggestion
```

ycsin commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -549,6 +549,12 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 	net_buf_add(buf, len);
 
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_ENSURE_SHORT_PKT_FOR_LAST_TX_PKT)) {
+		if (len != 0 && len % cdc_acm_get_bulk_mps(c_data) == 0) {
```
```suggestion
		if ((len != 0) && ((len % cdc_acm_get_bulk_mps(c_data)) == 0)) {
```

zeonchew replied:
Thanks @ycsin for the suggestions. Have pushed the change accordingly.

jfischer-no replied:
You should not follow this suggestion, please remove unnecessary parentheses.

ycsin replied:
> You should not follow this suggestion, please remove unnecessary parentheses.

Isn't that MISRA C rule 12.1 & 13.5?

zeonchew replied:
Hi @ycsin, @jfischer-no, looked around the coding style for `usbd` and it seems that the style is more towards minimizing parentheses. 
Hence, for now I think it's better to remove the parentheses for consistency within `usbd`. 
Let's move on with this, we would need the ZLP change to be merge in ASAP so that we could resolve this issue with windows USBSER driver.

ycsin replied:
🤷‍♂️ we can leave it for the misra people

ycsin commented on subsys/usb/device_next/class/Kconfig.cdc_acm:
```diff
@@ -21,6 +21,15 @@ config USBD_CDC_ACM_STACK_SIZE
+	help
+	  If the length of CDC ACM TX payload to be sent to USB Host
+	  is in multiple of max-packet-size of the BULK-IN endpoint,
+	  a Zero-Length Packet will be sent followig the last packet
```
```suggestion
	  a Zero-Length Packet will be sent following the last packet
```

zeonchew replied:
Hi @ycsin, thanks for catching that. pushed a fix for this.

ycsin reviewed (dismissed):
Small typo, LGTM otherwise

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -549,6 +549,12 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_ENSURE_SHORT_PKT_FOR_LAST_TX_PKT)) {
+		if ((len != 0) && ((len % cdc_acm_get_bulk_mps(c_data)) == 0)) {
+			udc_ep_buf_set_zlp(buf);
```
This will have a performance impact, because this forces ZLP even if new data is written to the ring buffer during the IN transfer. What should be done, in my opinion is enqueueing:
  * all the data if len < mps, OR
  * ROUND_DOWN(len, mps) if len >= mps

Then after the request is finished, if the ring buffer is empty and last transmitted buffer was mps multiple (which can be stored in some boolean flag), deliberately send ZLP.

zeonchew replied:
This is actually the idea of the initial code when this PR was raised. Which was kind of abandoned after the discussion that took place, to better utilize the existing ZLP flag for endpoints in UDC. 

Now I am wondering which path we should go here, as we are having different opinions on this: 
1. favor performance like what you have suggested
2. Simply go for the simpler handling that matches the existing handling. 
I personally favor performance in this regard.

tmon-nordic replied:
The ZLP can be added together with improved performance (even when the ZLP option is not selected). The existing handling has various other issues (e.g. #79104) and here it seems that the solution would be simplest if based on top of #79104.

zeonchew replied:
@tmon-nordic, It seems to be ideal for me that : 
1. #79104 use `ring_buffer_is_empty()` instead of `tx_fifo.altered` flags to check whether there are more transaction needed. Then, 
2. For this PR, we add one `ZLP needed` flag to indicate the current transmitting packet might need ZLP, and send ZLP out if tx_fifo is still empty by the time the current IN transaction completed. 

Shall we wait for @jfischer-no  input on these 2 items? 
@jfischer-no are you okay with the handling above for these 2 PR? If so I will make the changes accordingly, and we will be able to move forward. 

Thank you.

tmon-nordic replied:
I cannot really think of why this plan wouldn't work. I think you can just proceed to implementation without waiting for anyone's permission.

zeonchew replied:
Let me look for time to do the change and verification before pushing the change up.

zeonchew replied:
Hi @tmon-nordic, 
Changes discussed above made and tested working at my side. Please review again and see whether its all good. 
Thank you.

tmon-nordic replied:
This PR will inevitably conflict with #79104.

zeonchew replied:
Yeah @tmon-nordic , this will inevitably conflict with #79104. So I am planning to rebase either one of the PR once the other one make it into main branch. Is that the correct flow to go in this kind of scenario?

zeonchew replied:
Hi @tmon-nordic, Since #78160 (that replaces #79104) is merged in, I have resolved the merge conflict and pushed the change. Please have a look again and see whether its alright with you. 😉

tmon-nordic replied:
At High-Speed this already gives the best throughput because `cdc_acm_ep_pool` returns 512 bytes large buffers. At Full-Speed the `len = ring_buf_get(data->tx_fifo.rb, buf->data, buf->size);` would need to be changed according to https://github.com/zephyrproject-rtos/zephyr/pull/79193#discussion_r1787233847 to get the best throughput.

We might eventually get to reduce the `cdc_acm_ep_pool` buffer for Full-Speed only builds, so this can probably stay as is for now.

zeonchew replied:
IMO, from both of the options above, I think making cdc_acm_ep_pool size adjustable is the better option as it helps 
- Reduce memory footprint in the case where the buffer needed is only 64 bytes for Full-Speed only compilation. 
- If current shim/hal/controller for a chip could support autonomous transfer of multiple packets, it might give better performance as it could do the multiple packet transfer with lesser overhead. 

I suggest we keep it as is for now 😄

tmon-nordic commented on subsys/usb/device_next/class/usbd_cdc_acm.c:
```diff
@@ -117,6 +117,8 @@ struct cdc_acm_uart_data {
 	struct k_work rx_fifo_work;
 	atomic_t state;
 	struct k_sem notif_sem;
+	/* ZLP needed if no more tx is needed after this tx is done */
+	bool zlp_needed;
```
You can move it below `bool flow_ctrl;` to save some RAM on padding.

tmon-nordic replied:
Also `bool flow_ctrl;` could be moved around, so all the bool variables (`line_state_rts`, `line_state_dtr`, `flow_ctrl`, `zlp_needed`) would be next to each other, which would save few more bytes.

zeonchew replied:
Hi @tmon-nordic, I think that is a good point on the packing overhead. by moving both `bool flow_ctrl` and `bool zlp_needed` we could save 8 bytes of RAM, by moving them t the original padding space of `line_state_rts`, `line_state_dtr`. 
I have made the change as suggested.

tmon-nordic commented on subsys/usb/device_next/class/Kconfig.cdc_acm:
```diff
@@ -21,6 +21,16 @@ config USBD_CDC_ACM_STACK_SIZE
 	  USB CDC ACM workqueue stack size.
 
+config USBD_CDC_ACM_TX_ZLP_ENABLED
+	bool "USB CDC ACM transmission always end with short packet"
```
Arguably this could be potentially made `default y`. Windows driver is the original reason behind this, but Linux CDC ACM driver also can benefit from ZLP in corner case.

Linux CDC ACM driver will allocate URBs that can hold 2 Max Packet Size by default:
https://github.com/torvalds/linux/blob/c2ee9f594da826bea183ed14f2cc029c719bf4da/drivers/usb/class/cdc-acm.c#L1350-L1351C16
unless the `SINGLE_RX_URB` quirk is set.

If the CDC ACM device transmit data stream ends with packet that is exactly Max Packet Size bytes and ceases to send more data, on Linux host there is possibility that the last Max Packet Size packet will be withheld in (host side) URB until there is new data received from device. Depending on application protocol this may lead to stalls that are "fixed" with the `SINGLE_RX_URB` quirk.

The ZLP itself is completely invisible to higher level applications that use ttyACM interface (e.g. pyserial).

zeonchew replied:
Hi @tmon-nordic, Thanks for the information. I did not know about the `SINGLE_RX_URB` thing. It was set as default disabled as I was thinking that it is only meant for Windows Host Driver. Let's make this `default y` for now and see what the over reviewer thinks.  😄

jfischer-no replied:
I have no idea how MacOS handles this internally, but I tested it on MacOS Sonoma with pyserial, it seems to have no drawbacks. I would like to avoid this kind of always-on Kconfig option then and not introduce it at all.

ycsin commented:
ping @jfischer-no

zeonchew commented:
Hi @jfischer-no, please kindly review and see whether this PR is all good to you. Thank you.

jfischer-no reviewed (approved):
Thanks
